### ISSUE TITLE
Sample inference deployed using the modelcar functionality is not starting

### ISSUE THREAD

**Issue by bpradipt** [kserve/kserve#3506]
/kind bug

**What steps did you take and what happened:**
1. Deployed kserve release-0.12 using quick_install.sh
2. Edited inferenceservice-config to enable modelcar
3. Deployed a sample inferenceservice
```
apiVersion: "serving.kserve.io/v1beta1"
kind: "InferenceService"
metadata:
  name: "sklearn-iris"
spec:
  predictor:    
    model:
      modelFormat:
        name: sklearn
      storageUri: "oci://rhuss/sklearn-sample-model"
```

**What did you expect to happen:**

Successful inferenceservices url

**What's the InferenceService yaml:**
[To help us debug please run `kubectl get isvc $name -n $namespace -oyaml` and paste the output]

```
apiVersion: serving.kserve.io/v1beta1
kind: InferenceService
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"serving.kserve.io/v1beta1","kind":"InferenceService","metadata":{"annotations":{},"name":"sklearn-iris","namespace":"kserve-test"},"spec":{"predictor":{"model":{"modelFormat":{"name":"sklearn"},"storageUri":"oci://rhuss/sklearn-sample-model"}}}}
  creationTimestamp: "2024-03-10T07:43:07Z"
  finalizers:
  - inferenceservice.finalizers
  generation: 1
  name: sklearn-iris
  namespace: kserve-test
  resourceVersion: "34946804"
  uid: 8b38aa70-490a-48f4-97e8-a58243c550bc
spec:
  predictor:
    model:
      modelFormat:
        name: sklearn
      name: ""
      resources: {}
      storageUri: oci://rhuss/sklearn-sample-model
status:
  components:
    predictor:
      latestCreatedRevision: sklearn-iris-predictor-00001
  conditions:
  - lastTransitionTime: "2024-03-10T07:43:07Z"
    reason: PredictorConfigurationReady not ready
    severity: Info
    status: Unknown
    type: LatestDeploymentReady
  - lastTransitionTime: "2024-03-10T07:43:07Z"
    severity: Info
    status: Unknown
    type: PredictorConfigurationReady
  - lastTransitionTime: "2024-03-10T07:43:07Z"
    status: Unknown
    type: PredictorReady
  - lastTransitionTime: "2024-03-10T07:43:07Z"
    severity: Info
    status: Unknown
    type: PredictorRouteReady
  - lastTransitionTime: "2024-03-10T07:43:07Z"
    status: Unknown
    type: Ready
  - lastTransitionTime: "2024-03-10T07:43:07Z"
    reason: PredictorRouteReady not ready
    severity: Info
    status: Unknown
    type: RoutesReady
  modelStatus:
    states:
      activeModelState: ""
      targetModelState: Pending
    transitionStatus: InProgress
  observedGeneration: 1
```

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]
The following is the error in the deployment manifest
```
  message: 'Pod "sklearn-iris-predictor-00001-deployment-854449bb44-xj8lh" is invalid:
      [spec.volumes[2].name: Duplicate value: "kserve-provision-location", spec.containers[0].volumeMounts[2].mountPath:
      Invalid value: "/mnt": must be unique, spec.containers[3].name: Duplicate value:
      "modelcar"]'
```

**Environment:**

- Istio Version: 1.19.4
- Knative Version: knative-v1.10.1
- KServe Version: v0.12.0
- Kubeflow version: NA
- Cloud Environment:[k8s_istio/istio_dex/gcp_basic_auth/gcp_iap/aws/aws_cognito/ibm] aks
- Minikube/Kind version: NA
- Kubernetes version: (use `kubectl version`): v1.27.7
- OS (e.g. from `/etc/os-release`): Ubuntu 20.04

**Comment by surajssd** [comment]
Yes, can't really see what the generated deployment is, because the one that shows up when querying k8s is half-baked.

**Comment by rhuss** [comment]
Thanks for the bug report! I just could replicate it. Now on it .... :-)

**Comment by rhuss** [comment]
It's indeed a bug because the modelcar's mutating webhook code was not idempotent, and the WebHook introduced an `IfNeeded` reinvocation policy with https://github.com/kserve/kserve/commit/39b8a6748732349f72d0701076c85e66d630755f (btw, @israel-hdez why was this needed in the context of adding compatibility with the Istion CNI plugin? I am asking because it (a) introduces quite some overhead in calling the webhook code multiple times until it converges (~around 7 to 8 times, what I've seen~ after the fix, and the method is only called twice. The excess calls happen only when K8s tries to reconcile again in the error case. So the overhead is smaller) and (b) can trigger issues like these when the code is not idempotent).

In this case, the webhook's modelcar part unconditionally adds an `emptyDir` volume to the list of volumes without checking whether it has already been added.

Going to fix this now, stay tuned ...

**Comment by rhuss** [comment]
BTW, I also have the feeling that the rotation code for the serviceaccount secrets from Kubernetes does not work well with `IfNeeded` since I also see several subsequent calls to the webhook with only these diffs (all within one second) in the `AdmissionRequest`:

```diff
         "manager": "kube-controller-manager",
         "operation": "Update",
         "apiVersion": "v1",
-        "time": "2024-03-13T06:41:09Z",
+        "time": "2024-03-13T06:43:09Z",
         "fieldsType": "FieldsV1",
         "fieldsV1": {
           "f:metadata": {
@@ -395,7 +395,7 @@
   "spec": {
     "volumes": [
       {
-        "name": "kube-api-access-hgs44",
+        "name": "kube-api-access-fznsf",
         "projected": {
           "sources": [
             {
@@ -479,7 +479,7 @@
         },
         "volumeMounts": [
           {
-            "name": "kube-api-access-hgs44",
+            "name": "kube-api-access-fznsf",
             "readOnly": true,
             "mountPath": "/var/run/secrets/kubernetes.io/serviceaccount"
           }
@@ -670,7 +670,7 @@
         },
         "volumeMounts": [
           {
-            "name": "kube-api-access-hgs44",
+            "name": "kube-api-access-fznsf",
             "readOnly": true,
             "mountPath": "/var/run/secrets/kubernetes.io/serviceaccount"
           }
```

But that's another story ....

**Comment by rhuss** [comment]
A fix has been added in #3517; @bpradipt thanks for reporting this!

**Comment by israel-hdez** [comment]
> btw, @israel-hdez why was this needed in the context of adding compatibility with the Istion CNI plugin? I am asking because it (a) introduces quite some overhead in calling the webhook code multiple times

AFAIK, the Istio sidecar is injected via a mutating webhook. Since webhooks run in undefined order, there is the chance that KServe one runs first. If KServe webhook runs first, it would not set the storage-initializer security context to work fine with Istio with the CNI plugin, because it would determine it is not in the mesh (because of the missing not yet injected sidecar).

So, `IfNeeded` was needed to let the KServe webhook "see" if a sidecar is injected at some point and properly set the storage-initializer security context accordingly.

**Comment by rhuss** [comment]
@israel-hdez thanks for the clarification, makes sense to me. It's more robust anyways to stay idempotent, so the fix in #3517 is not harmful anyhow.
